### PR TITLE
Verify the construction of SubstitutionMaps

### PR DESCRIPTION
**Pull request by swiftix**
After substitution maps are constructed, check their invariants. It helps to catch very subtle bugs related to substitutions.

### CODE REVIEW

**Comment by swiftix** [comment]
@slavapestov Looks OK?

File: lib/AST/SubstitutionMap.cpp [hunk 1]
```diff
@@ -395,6 +398,40 @@ SubstitutionMap::combineSubstitutionMaps(const SubstitutionMap &firstSubMap,
     });
 }
 
+void SubstitutionMap::verify() const {
+#ifndef NDEBUG
```
Comment (slavapestov) [inline lib/AST/SubstitutionMap.cpp]: Maybe the entire method should be wrapped in the #ifndef? What do you think @jckarter

**Comment by slavapestov** [comment]
Looks good, thanks!

File: lib/AST/SubstitutionMap.cpp (hunk 1 above)
Comment (swiftix) [inline lib/AST/SubstitutionMap.cpp]: If the whole method is inside #ifndef, then all call-sites of this methods should be inside #ifndef as well. I find it a bit ugly.

Comment (slavapestov) [inline lib/AST/SubstitutionMap.cpp]: Right, but then it will not add any overhead at all in no-assert builds.

Comment (swiftix) [inline lib/AST/SubstitutionMap.cpp]: Well, the overhead of calling an empty method should not affect the performance IMHO.

Comment (jckarter) [inline lib/AST/SubstitutionMap.cpp]: This seems like a good opportunity to introduce an -enable-verifiers flag so that they're runtime-enabled rather than compiled out of release builds.

Comment (swiftix) [inline lib/AST/SubstitutionMap.cpp]: @jckarter I agree with you. But introducing an -enable-verifiers flag should be a separate commit, which would introduce a flag and update all the places where we currently compile out the code in release builds.

**Comment by swiftix** [comment]
@swift-ci please test

**Comment by swiftix** [comment]
@swift-ci please test

**Comment by swiftix** [comment]
@swift-ci please test

**Comment by swiftix** [comment]
@swift-ci please test

**Comment by slavapestov** [comment]
@swiftix Next time can you use a better name for your PR branch than 'master'? :)

**Comment by swiftix** [comment]
@slavapestov Sure. Usually I do. This time it came as a quick fix and I forgot to create a branch ;-)
